libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -1951,53 +1947,106 @@ void GCS::send_textv(MAV_SEVERITY severity, const char *fmt, va_list arg_list, u
     if (notify) {
         notify->send_text(first_piece_of_text);
     }
+
+    // push the messages out straight away until the vehicle states
+    // that it is initialised.  At that point we can assume
+    // update_send is being called
+    if (!vehicle_initialised()) {
+        service_statustext();
+    }
 }
 
-/*
-    send a statustext message to specific MAVLink connections in a bitmask
- */
 void GCS::service_statustext(void)
 {
-    // create bitmask of what mavlink ports we should send this text to.
-    // note, if sending to all ports, we only need to store the bitmask for each and the string only once.
-    // once we send over a link, clear the port but other busy ports bit may stay allowing for faster links
-    // to clear the bit and send quickly but slower links to still store the string. Regardless of mixed
-    // bitrates of ports, a maximum of _status_capacity strings can be buffered. Downside
-    // is if you have a super slow link mixed with a faster port, if there are _status_capacity
-    // strings in the slow queue then the next item can not be queued for the faster link
+    WITH_SEMAPHORE(_statustext_queue.semaphore());
 
     if (_statustext_queue.is_empty()) {
         // nothing to do
         return;
     }
 
-    for (uint8_t idx=0; idx<_status_capacity; ) {
-        statustext_t *statustext = _statustext_queue[idx];
+    for (uint8_t i=first_backend_to_send; i<num_gcs(); i++) {
+        chan(i)->service_statustext();
+    }
+    for (uint8_t i=0; i<first_backend_to_send; i++) {
+        chan(i)->service_statustext();
+    }
+
+    _statustext_queue.prune();
+}
+
+void GCS::StatusTextQueue::prune(void)
+{
+    // consider pruning the statustext queue of ancient entries
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - last_prune_ms < 1000) {
+        return;
+    }
+    last_prune_ms = now_ms;
+
+    const uint16_t now16_ms = AP_HAL::millis16();
+    for (uint8_t idx=0; idx<available(); ) {
+        const GCS::statustext_t *statustext = (*this)[idx];
         if (statustext == nullptr) {
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
             break;
         }
+        if (now16_ms - statustext->entry_created_ms > 5000) {
```

Comment:
it is not safe to subtract uint16_t without a cast, wrap won't work correctly. It is best to put the difference into an intermediate variable

Reply:
Fixed.